OSSM-9884: Create egress gateway using gateway API (#95506)
openshift/openshift-docs: modules/ossm-directing-egress-traffic-through-a-gateway-using-kubernetes-gateway-api.adoc

// This procedure is used in the following assembly:
// * service-mesh-docs-main/gateways/ossm-directing-outbound-traffic-through-a-gateway

:_mod-docs-content-type: PROCEDURE
[id="ossm-directing-egress-traffic-through-a-gateway-using-kubernetes-gateway-api_{context}"]
= Directing egress traffic through a gateway by using the {k8s} Gateway API

Use the {k8s} Gateway API to direct outbound HTTP traffic through an egress gateway.

.Prerequisites

* You installed an {istio} control plane.

* You configured the `Istio` and `IstioCNI` resources. 

.Procedure

. Optional: Enable the {k8} Gateway API custom resource definitions (CRDs). 
+
[NOTE]
====
As of {k8s} 1.28 and {ocp-product-title} 4.18 or earlier version of {product-title}, the {k8s} Gateway API CRDs are not available by default and you must enabled the CRDs before you can use them. {ocp-product-title} 4.19 and later versions enable the CRDs by default.
====

.. Create a YAML file named `gateway-cr` that enables the Kubernetes Gateway API custom resource definitions (CRDs). 
+

> .Example {k8s} Gateway CR file

Author: done

[source,yaml,subs="attributes,verbatim"]
----
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: bookinfo-gateway
spec:
  gatewayClassName: istio
  listeners:
  - name: http
    port: 80
    protocol: HTTP
    allowedRoutes:
      namespaces:
        from: Same
---
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: bookinfo
spec:
  parentRefs:
  - name: bookinfo-gateway
  rules:
  - matches:
    - path:
        type: Exact
        value: /productpage
    - path:
        type: PathPrefix
        value: /static
    - path:
        type: Exact
        value: /login
    - path:
        type: Exact
        value: /logout
    - path:
        type: PathPrefix
        value: /api/v1/products
    backendRefs:
    - name: productpage
      port: 9080
----

.. Apply the YAML file by running the following command:
+
[source,terminal]
----

> oc apply -f gateway-cr.yml

Author: done

  listeners:
  - name: http
    hostname: httpbin.org
    port: 80
    protocol: HTTP
    allowedRoutes:
      namespaces:
        from: Same
---
# HTTPRoute to direct traffic from sidecars to the egress gateway
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: direct-httpbin-to-egress-gateway
spec:
  parentRefs:
  - kind: ServiceEntry
    group: networking.istio.io
    name: httpbin-ext
  rules:
  - backendRefs:
    - name: httpbin-egress-gateway-istio
      port: 80
---
# HTTPRoute to forward traffic from the egress gateway to httpbin.org
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: forward-httpbin-from-egress-gateway
spec:
  parentRefs:
  - name: httpbin-egress-gateway
  hostnames:
  - httpbin.org
  rules:
  - backendRefs:
    - kind: Hostname
      group: networking.istio.io
      name: httpbin.org
      port: 80
----

.. Apply the YAML file by running the following command:
+
[source,terminal]
----
oc apply -f egress-gateway-cr.yml
----

.Verification

. Verify the status of the gateway configuration by running the following command:
+
[source,terminal]
----
$ oc describe gateway -n egress-gateway
----
+

> Desired output is indicated  by `Programmed` showing in the `Status` column.

Author: done

// This procedure is used in the following assembly:
// * service-mesh-docs-main/gateways/ossm-directing-outbound-traffic-through-a-gateway

:_mod-docs-content-type: PROCEDURE
[id="ossm-directing-egress-traffic-through-a-gateway-using-kubernetes-gateway-api_{context}"]
= Directing egress traffic through a gateway by using the {k8s} Gateway API

Use the {k8s} Gateway API to direct outbound HTTP traffic through an egress gateway.

.Prerequisites

> * You installed an {istio} control plane and configured the `Istio` and `IstioCNI` resources 

Author: Done

.Procedure

> . Optional: Enable the Kubernetes Gateway API CRDs by running the following command:

Reviewer: ```suggestion
. Optional: Enable the Kubernetes Gateway API custom resource definitions (CRDs) by running the following command:
```

I don't see CRD fully spelled out in the document.

+
[source,terminal]
----
$ oc apply -f https://raw.githubusercontent.com/istio/istio/master/samples/bookinfo/gateway-api/bookinfo-gateway.yaml
----
+
[NOTE]
====

> As of {k8s} 1.28 and {ocp-product-title} 4.18 or earlier, the Kubernetes Gateway API CRDs are not available by default and must be enabled to be used. the CRDs are enabled by default in  {ocp-product-title} 4.19 and later.

Author: done

====

> . Create a namespace called egress-gateway by running the following command:

Reviewer: ```suggestion
. Create a namespace called `egress-gateway` by running the following command:
```

Author: Thanks! Done.

+
[source,terminal]
----
$ oc create namespace egress-gateway
----

> . Apply the istio-injection label to the namespace by running the following command:

Reviewer: ```suggestion
. Apply the `istio-injection` label to the namespace by running the following command:
```

Author: Thanks! Done.

+
[source,terminal]
----
$ oc label namespace egress-gateway istio-injection=enabled
----

. Apply the example egress gateway configuration by running the following command:
+
[source,terminal]
----

> $ oc apply -f https://raw.githubusercontent.com/istio-ecosystem/sail-operator/main/chart/samples/egress-gateway-gw-api.yaml -n egress-gateway

Reviewer: Would hardcoding the YAML be safer?

Author: Done.

----

> . Verify the status of the gateway configuration by running the following command:

Author: Done

+
[source,terminal]
----
$ oc describe gateway -n egress-gateway
----
+

> `Programmed` should appear in the `Status` column.

Author: Done.

. Create a `curl` pod in the `egress-gateway` namespace by running the following command:
+
[source,terminal]
----
$ oc run test-pod --image=curlimages/curl:latest -n egress-gateway --rm -it --restart=Never -- sh
----

> . Using the `curl` client, verify that you can access `httpbin.org` through the egress gateway. Run the following command:

Reviewer: ```suggestion
. By using the `curl` client, verify that you can access `httpbin.org` through the egress gateway by entering following command:
```

+
[source,terminal]
----
$ curl -v http://httpbin.org/get
----
+

> You should see a response from `httpbin.org` indicating that egress traffic is being routed through the configured gateway.

Author: Done
